components/tikv_util/src/sys/cgroup.rs

```diff
@@ -1,520 +1,236 @@
-const CPU_PERIOD: &str = "cpu.cfs_period_us";
-const CPUSET_SUBSYS: &str = "cpuset";
-const CPUSET_CPUS: &str = "cpuset.cpus";
-
-const MOUNTINFO_SEP: &str = " ";
-const OPTIONS_SEP: &str = ",";
-const OPTIONAL_FIELDS_SEP: &str = "-";
-
-const CGROUP_SEP: &str = ":";
-const SUBSYS_SEP: &str = ",";
-
-#[derive(Debug, Error)]
-pub enum Error {
-    #[error("{0}")]
-    Other(#[from] Box<dyn error::Error + Sync + Send>),
-    #[error("{0}")]
-    Io(#[from] std::io::Error),
-    #[error("{0}")]
-    PathErr(#[from] std::path::StripPrefixError),
-}
-
-enum MountInfoFieldPart1 {
-    MountID = 0,
-    ParentID,
-    DeviceID,
-    Root,
-    MountPoint,
-    Options,
-    OptionalFields,
-
-    Count,
-}
-
-enum MountInfoFieldPart2 {
-    FSType = 0,
-    MountSource,
-    SuperOptions,
-
-    Count,
-}
-
-#[derive(Debug, PartialEq)]
-pub struct MountPoint {
-    pub mount_id: i32,
-    pub parent_id: i32,
-    pub device_id: String,
-    pub root: String,
-    pub mount_point: String,
-    pub options: Vec<String>,
-    pub optional_fields: Vec<String>,
-    pub fs_type: String,
-    pub mount_source: String,
-    pub super_options: Vec<String>,
+pub struct CGroupSys {
+    cgroups: HashMap<String, String>,
+    is_v2: bool,
 }
 
-impl MountPoint {
-    // Converts an absolute path inside the *MountPoint's file system to
-    // the host file system path in the mount namespace the *MountPoint belongs to.
-    pub fn translate(&self, abs_path: &str) -> Result<String, Error> {
-        let abs = Path::new(abs_path);
-        let rel = abs.strip_prefix(&self.root)?;
-        let path = Path::new(&self.mount_point).join(rel);
-        if let Some(path) = path.to_str() {
-            Ok(String::from(path))
+impl CGroupSys {
+    pub fn new() -> Self {
+        let mut f = File::open(&format!("/proc/{}/cgroup", process::id())).unwrap();
+        let mut lines = String::new();
+        f.read_to_string(&mut lines).unwrap();
+
+        let is_v2 = is_cgroup2_unified_mode();
+        let cgroups = if !is_v2 {
+            parse_proc_cgroup_v1(&lines)
         } else {
-            Err(box_err!("empty path"))
-        }
-    }
-}
-
-pub fn parse_mount_point_from_line(line: &str) -> Result<MountPoint, String> {
-    let fields: Vec<String> = line.split(MOUNTINFO_SEP).map(|s| s.to_string()).collect();
-    if fields.len() < MountInfoFieldPart1::Count as usize + MountInfoFieldPart2::Count as usize {
-        return Err("mount point format invalid".to_string());
+            parse_proc_cgroup_v2(&lines)
+        };
+        CGroupSys { cgroups, is_v2 }
     }
 
-    let mut sep_pos = MountInfoFieldPart1::OptionalFields as usize;
-    let mut found_sep = false;
-    for field in &fields[MountInfoFieldPart1::OptionalFields as usize..] {
-        if field == OPTIONAL_FIELDS_SEP {
-            found_sep = true;
-            break;
+    /// -1 means no limit.
+    pub fn memory_limit_in_bytes(&self) -> i64 {
+        if self.is_v2 {
```

Comment:
How about adding some comments explaining the format of files and difference between versions?

Reply:
Now I add a comment sector.